homeassistant/helpers/script.py

```diff
@@ -47,107 +49,126 @@ def __init__(self, hass: HomeAssistant, sequence, name: str=None,
+                            delay.async_render())
+
+                self._async_unsub_delay_listener = \
+                    async_track_point_in_utc_time(
                         self.hass, script_delay,
                         date_util.utcnow() + delay)
-                    self._cur = cur + 1
-                    if self._change_listener:
-                        self._change_listener()
-                    return
+                self._cur = cur + 1
+                self._trigger_change_listener()
+                return
 
-                elif CONF_CONDITION in action:
-                    if not self._check_condition(action, variables):
-                        break
+            elif CONF_CONDITION in action:
+                if not self._async_check_condition(action, variables):
+                    break
 
-                elif CONF_EVENT in action:
-                    self._fire_event(action)
+            elif CONF_EVENT in action:
+                self._async_fire_event(action)
 
-                else:
-                    self._call_service(action, variables)
+            else:
+                yield from self._async_call_service(action, variables)
 
-            self._cur = -1
-            self.last_action = None
-            if self._change_listener:
-                self._change_listener()
+        self._cur = -1
+        self.last_action = None
+        self._trigger_change_listener()
 
     def stop(self) -> None:
         """Stop running script."""
-        with self._lock:
-            if self._cur == -1:
-                return
+        run_callback_threadsafe(self.hass.loop, self.async_stop).result()
 
-            self._cur = -1
-            self._remove_listener()
-            if self._change_listener:
-                self._change_listener()
+    def async_stop(self) -> None:
+        """Stop running script."""
+        if self._cur == -1:
+            return
 
-    def _call_service(self, action, variables):
+        self._cur = -1
+        self._async_remove_listener()
+        self._trigger_change_listener()
+
+    @asyncio.coroutine
+    def _async_call_service(self, action, variables):
         """Call the service specified in the action."""
         self.last_action = action.get(CONF_ALIAS, 'call service')
         self._log("Executing step %s" % self.last_action)
-        service.call_from_config(self.hass, action, True, variables,
-                                 validate_config=False)
+        yield from service.async_call_from_config(
+            self.hass, action, True, variables, validate_config=False)
 
-    def _fire_event(self, action):
+    def _async_fire_event(self, action):
         """Fire an event."""
         self.last_action = action.get(CONF_ALIAS, action[CONF_EVENT])
         self._log("Executing step %s" % self.last_action)
-        self.hass.bus.fire(action[CONF_EVENT], action.get(CONF_EVENT_DATA))
+        self.hass.bus.async_fire(action[CONF_EVENT],
```

Comment:
Because we are guaranteed that everything under firing an event will not `yield from` anything. We might encounter coroutines but add them to the work queue.

Less coroutines = better 👍